data/data-pipeline/data_pipeline/etl/sources/cdc_life_expectancy/etl.py

```diff
@@ -1,58 +1,137 @@
 
 logger = get_module_logger(__name__)
 
 
 class CDCLifeExpectancy(ExtractTransformLoad):
     def __init__(self):
-        self.FILE_URL: str = "https://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NVSS/USALEEP/CSV/US_A.CSV"
+        self.GEO_LEVEL = ValidGeoLevel.CENSUS_TRACT
+        self.PUERTO_RICO_EXPECTED_IN_DATA = False
+
+        self.USA_FILE_URL: str = "https://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NVSS/USALEEP/CSV/US_A.CSV"
+
+        self.STATES_MISSING_FROM_USA_FILE = ["23", "55"]
+
+        # For some reason, LEEP does not include Maine or Wisconsin in its "All of
+        # USA" file. Load these separately.
+        self.WISCONSIN_FILE_URL: str = "https://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NVSS/USALEEP/CSV/WI_A.CSV"
+        self.MAINE_FILE_URL: str = "https://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NVSS/USALEEP/CSV/ME_A.CSV"
+
         self.OUTPUT_PATH: Path = (
             self.DATA_PATH / "dataset" / "cdc_life_expectancy"
         )
 
         self.TRACT_INPUT_COLUMN_NAME = "Tract ID"
-        self.LIFE_EXPECTANCY_FIELD_NAME = "Life expectancy (years)"
+        self.STATE_INPUT_COLUMN_NAME = "STATE2KX"
 
         # Constants for output
         self.COLUMNS_TO_KEEP = [
             self.GEOID_TRACT_FIELD_NAME,
-            self.LIFE_EXPECTANCY_FIELD_NAME,
+            field_names.LIFE_EXPECTANCY_FIELD,
         ]
 
         self.raw_df: pd.DataFrame
         self.output_df: pd.DataFrame
 
-    def extract(self) -> None:
-        logger.info("Starting data download.")
-
-        download_file_name = (
-            self.get_tmp_path() / "cdc_life_expectancy" / "usa.csv"
-        )
+    def _download_and_prep_data(
+        self, file_url: str, download_file_name: pathlib.Path
+    ) -> pd.DataFrame:
         download_file_from_url(
-            file_url=self.FILE_URL,
+            file_url=file_url,
             download_file_name=download_file_name,
             verify=True,
         )
 
-        self.raw_df = pd.read_csv(
+        df = pd.read_csv(
             filepath_or_buffer=download_file_name,
             dtype={
                 # The following need to remain as strings for all of their digits, not get converted to numbers.
                 self.TRACT_INPUT_COLUMN_NAME: "string",
+                self.STATE_INPUT_COLUMN_NAME: "string",
             },
             low_memory=False,
         )
 
+        return df
+
+    def extract(self) -> None:
+        logger.info("Starting data download.")
+
+        all_usa_raw_df = self._download_and_prep_data(
+            file_url=self.USA_FILE_URL,
+            download_file_name=self.get_tmp_path()
+            / "cdc_life_expectancy"
+            / "usa.csv",
+        )
+
+        # Check which states are missing
+        states_in_life_expectancy_usa_file = list(
+            all_usa_raw_df[self.STATE_INPUT_COLUMN_NAME].unique()
+        )
+
+        # Expect that PR, Island Areas, and Maine/Wisconsin are missing
```

Comment:
I can't decide if this is over-engineering or not, but given the missing codes are a bug anyway, it _almost_ feels worth doing the extra work to just look at the states we got back and download any missing ones but inferring them, since the URLs can be constructed pretty easily. Then, if ME and WI come back, we'll just stop doing the extra work, and if they drop MD randomly, we would automatically recover from that too. But, I also get that that hides some weird stuff happening on a data partner from us and maybe we want to explicitly fail if anything changes on their side...

Reply:
Yeah - I get what you're saying. But I think I'd prefer to be explicit and fail fast rather than engineer something that probably but not 100% definitely works to fully recover. 

I mainly put this code in because @BethMattern is gonna send an email to the CDC asking them to fix this source data, so I wanted us to have a heads-up when it changes.

Reply:
Works for me.